README
Acme::Time::Constant



INSTALLATION

This is a Perl module distribution. It should be installed with whichever
tool you use to manage your installation of Perl, e.g. any of

  cpanm .
  cpan  .
  cpanp -i .

Consult http://www.cpan.org/modules/INSTALL.html for further instruction.
Should you wish to install this module manually, the procedure is

  perl Makefile.PL
  make
  make test
  make install

COPYRIGHT AND LICENSE

This software is copyright (c) 2016 by Kent Fredric
<kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Release history for Acme-Time-Constant

0.001007 2017-03-07T08:14:06Z 1dbeee7

0.001006 2017-03-07T08:12:13Z 805d720
 - Bugfix: Avoid test failures with -Ddefault_inc_excludes_dot

 [Dependencies::Stats]
 - Dependencies changed since 0.001005, see misc/*.deps* for details
 - develop: +1 ↑1 (suggests: ↑2)
 - test: ↓1

0.001005 2016-04-27T14:50:17Z adc5064
 - Requiring the newest Time::Warp, which appears to have avoided the nvsize bug reported in issue #1
 - Alternatively: I blame CERN

 [Dependencies::Stats]
 - Dependencies changed since 0.001004, see misc/*.deps* for details
 - develop: +4 ↑1 (suggests: ↑1)
 - runtime: ↑1

0.001004 2015-06-23T04:50:45Z f351e75
 [00 Minor]
 - Attempt 5.6 support
 - This will come in handy when this module supports time travel and has to be used back before 5.8 exists without
   breaking timespace.

 [Dependencies::Stats]
 - Dependencies changed since 0.001003, see misc/*.deps* for details
 - configure: +1 (recommends: ↑1)
 - develop: +9 ↑3 -3 (recommends: +1, suggests: ↑2)
 - runtime: ↓1 -1
 - test: +1 (recommends: ↑1 ↓1)

0.001003 2014-08-30T07:49:40Z
 [00 Trivial]
 - Improve test coverage.
 - Pull timespace and subspace closer together.

0.001002 2014-08-30T00:35:43Z
 [00 Trivial]
 - no code changes
 - metadata

 [Dependencies::Stats]
 - Dependencies changed since 0.001001, see misc/*.deps* for details
 - develop: +1 ↑1 (suggests: ↑2)

 [Metadata]
 - Fixup author

0.001001 2014-08-15T21:11:15Z
 [00 Trivial]
 - CPANDAY!
 - no code changes.
 - whitespace adjustment.

 [Dependencies::Stats]
 - Dependencies changed since 0.001000, see misc/*.deps* for details
 - develop: +1 (suggests: ↑1)

 [Misc]
 - Whitespace moved down by improving pre-build source.

0.001000 2014-08-04T04:27:24Z
 - First version.


lib/Acme/Time/Constant.pm
use 5.006;    # Our
use strict;
use warnings;

package Acme::Time::Constant;

our $VERSION = '0.001007';

# ABSTRACT: Run any code in constant time.

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

use Sub::Exporter::Progressive -setup => { exports => [qw( constant_time )] };

use Time::HiRes qw( gettimeofday );
use Time::Warp qw( to time );

my $old_import;
BEGIN { $old_import = \&import }

{
  ## no critic (TestingAndDebugging::ProhibitNoWarnings)
  no warnings 'redefine';

  sub import {
    *CORE::GLOBAL::time = *Time::Warp::time;
    goto $old_import;
  }
}








sub constant_time {
  my $nargs = ( my ( $time, $callback ) = @_ );

  if ( $nargs < 2 ) {
    $callback = $time;
    $time     = 1;
  }
  my $now = time;
  $callback->();
  to( $now + $time );
  return;
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Acme::Time::Constant - Run any code in constant time.

=head1 VERSION

version 0.001007

=head1 SYNOPSIS

This code contains within it, the golden calf of computer science: The ability to execute any code in constant time.

  use Acme::Time::Constant qw( constant_time );

  my $timestamp = time;

  constant_time( '1' => sub {
    for ( 0 .. 10 ) { }
  } );

  print $timestamp - time;  # 1 second

  constant_time( '1' => sub {
    for ( 0 .. 10_000_000 ) { }
  } );

  print $timestamp - time; # 2 seconds

=head1 BUGS

=head4 General relativity does not hold in the context of this code.

As such, observed time may differ greatly from the time relative to the execution of this code.

B<Workaround:> Get closer to the speed of light.

=head4 Measurements may not be exactly constant.

But Big O is OK with this.  We care not that X takes 1.1 seconds and Y takes 1.2 seconds, as long as the variation
is not subject to the size of Y or X.

Random variation between 0.5 and 1.5 seconds is thus within the range of "constant".

B<Workaround:> Imbibe a minimum of 1 L<< Litre|https://en.wikipedia.org/wiki/Litre >> of your favourite neurotoxic substance
before attempting to code.

=head4 Time::HiRes cannot be trusted.

C<Time::HiRes> is under the influence of the Illuminati and as such is part of a conspiracy to prevent us from experiencing
supernormal time.

Using C<Time::HiRes> in your code will subsequently give the illusion that the code no longer executes in constant time.

It is wrong. The constant time is simply functioning on an alternative timeline which the Illuminati seek to repress
knowledge of.

One may note that this module depends on C<Time::HiRes>, but this is simply our devilish tactic to make the conspirators
think we're willingly playing for them. You know better. ;)

=head1 FUNCTIONS

=head2 C<constant_time>

  constant_time( $time, $callback );
  constant_time( $callback ); # $time defaults to 1 second.

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Kent Fredric <kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


maint/perlcritic.rc.gen.pl
#!/usr/bin/env perl
## no critic (Modules::RequireVersionVar)

# ABSTRACT: Write an INI file from a bundle

use 5.008;    # utf8
use strict;
use warnings;
use utf8;

our $VERSION = 0.001;

use Carp qw( croak carp );
use Perl::Critic::ProfileCompiler::Util qw( create_bundle );
use Path::Tiny qw(path);

## no critic (ErrorHandling::RequireUseOfExceptions)
my $bundle = create_bundle('Example::Author::KENTNL');
$bundle->configure;

my @stopwords = (
  qw(
  neurotoxic supernormal timeline
    )
);
for my $wordlist (@stopwords) {
  $bundle->add_or_append_policy_field( 'Documentation::PodSpelling' => ( 'stop_words' => $wordlist ) );
}
$bundle->add_or_replace_policy_field('Documentation::PodSpelling' => ( 'spell_command' => 'aspell list --lang en_GB' ));

#$bundle->add_or_append_policy_field(
#  'Subroutines::ProhibitCallsToUndeclaredSubs' => ( 'exempt_subs' => 'String::Formatter::str_rf' ), );

$bundle->remove_policy('ErrorHandling::RequireUseOfExceptions');

#$bundle->remove_policy('CodeLayout::RequireUseUTF8');

#$bundle->remove_policy('ErrorHandling::RequireCarping');
#$bundle->remove_policy('NamingConventions::Capitalization');

my $inf = $bundle->actionlist->get_inflated;

my $config = $inf->apply_config;

{
  my $rcfile = path('./perlcritic.rc')->openw_utf8;
  $rcfile->print( $config->as_ini, "\n" );
  close $rcfile or croak 'Something fubared closing perlcritic.rc';
}
my $deps = $inf->own_deps;
{
  my $target = path('./misc');
  $target->mkpath if not $target->is_dir;

  my $depsfile = $target->child('perlcritic.deps')->openw_utf8;
  for my $key ( sort keys %{$deps} ) {
    $depsfile->printf( "%s~%s\n", $key, $deps->{$key} );
    *STDERR->printf( "%s => %s\n", $key, $deps->{$key} );
  }
  close $depsfile or carp 'Something fubared closing perlcritic.deps';
}



Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.009.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Run any code in constant time.",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Acme-Time-Constant",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "Acme::Time::Constant",
  "PREREQ_PM" => {
    "Sub::Exporter::Progressive" => 0,
    "Time::HiRes" => 0,
    "Time::Warp" => "0.52",
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::More" => 0
  },
  "VERSION" => "0.001007",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t"
  }
);


my %FallbackPrereqs = (
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Sub::Exporter::Progressive" => 0,
  "Test::More" => 0,
  "Time::HiRes" => 0,
  "Time::Warp" => "0.52",
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
